hw/chisel/doc/xdma.md

```diff
@@ -0,0 +1,238 @@
+
+## Features
+
+This section introduces three key features of XDMA:
+
+### New Data-fetching Unit (Streamer) to Support Powerful Direct Memory Access
+
+The TCDM interconnect at the top of the figure provides full-region data access with a one-cycle delay if there is no contention. XDMA's configuration includes variable bit widths for both memory and AXI channels. The number of data streamers on both reader and writer sides is automatically determined by these configurations, e.g.,:
+
+- **TCDMDataWidth=64** and **AXIDataWidth=512**: Typical for Snax clusters. XDMA will instantiate eight channels for both reading and writing at the TCDM side, concatenating them into a single 512-bit channel at the AXI side.
+- **TCDMDataWidth=64** and **AXIDataWidth=64**: Typical for global scratchpad memory/SRAM. XDMA will instantiate only one channel on each side, adopting a simpler architecture used by most other DMAs.
+
+Each TCDM channel has its own address and data caches, enabling independent data fetching with minimal interference from other channels. Channels can be selectively disabled to save energy and simplify memory initialization (e.g., turning off all reader channels and enabling writer channels to write zeros to memory). The byte mask is also implemented for byte-level granularity in writing.
+
+Data from the streamer is concatenated into a wide bus and forwarded to the datapath extension.
+
+### Extendable Datapath Extension to Enable *Computation on Copy* Capability
+
+The Datapath Extension is a critical component allowing on-the-fly data processing such as matrix transposing, max-pooling, average-pooling, and data stream compression/decompression. By supplying the necessary parameters to the XDMA generator, extension modules are instantiated sequentially, with each module's output feeding into the next module's input.
+
+Designing a custom datapath extension is straightforward thanks to Chisel's flexibility. The pipeline is automatically inserted, CSR configs are automatically wired, infrastructure for enabling/disabling extensions is added, and the corresponding C header for runtime is automatically generated. Integrating native extensions written in Chisel or extensions written in SystemVerilog is remarkably simple.
+
+### Distributed Datapath for Cluster-to-Cluster or Host-to-Cluster Memory Access
+
+XDMA can perform data copying within the same memory region or across different regions. If the source and destination addresses are from the same region, data is looped back directly without utilizing the AXI, conserving energy. If addresses span different regions, the data is handled in three steps:
+1. The controller recognizes that the source data is from a non-local region and flattens the configuration into a one-dimensional frame.
+2. The configuration frame is sent to the remote XDMA to request the data.
+3. The data is relayed back from the remote to the local site.
```

Comment:
Can you request data from a remote XDMA or only send data to a remote XDMA or both?

Reply:
We can only request the data from the remote. Write to TCDM only happens at the local. Choose not to support both condition is because the synchronization can be a problem, and it potentially become the cause of deadlock.